File: discipline_android_primitive_ts/chronic/duration.ts
import { Branded, TextualError, Tried } from "../x.ts";

const brand = Symbol();

export type Duration = Branded<typeof brand, number>;

export const construct = (milliseconds: number): Duration => {
  return Branded(brand, milliseconds);
};

export const MAXIMUM_MILLISECONDS = Number.MAX_SAFE_INTEGER;
export const MAXIMUM_SECONDS = Math.floor(MAXIMUM_MILLISECONDS / 1000);
export const MAXIMUM_MINUTES = Math.floor(MAXIMUM_SECONDS / 60);
export const MAXIMUM_HOURS = Math.floor(MAXIMUM_MINUTES / 60);
export const MAXIMUM_DAYS = Math.floor(MAXIMUM_HOURS / 24);
export const MAXIMUM_WEEKS = Math.floor(MAXIMUM_DAYS / 7);

export const MILLISECONDS_PER_SECOND = 1000;
export const MILLISECONDS_PER_MINUTE = MILLISECONDS_PER_SECOND * 60;
export const MILLISECONDS_PER_HOUR = MILLISECONDS_PER_MINUTE * 60;
export const MILLISECONDS_PER_DAY = MILLISECONDS_PER_HOUR * 24;
export const MILLISECONDS_PER_WEEK = MILLISECONDS_PER_DAY * 7;

export const fromMilliseconds = (milliseconds: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(milliseconds)) {
    const it = TextualError.create("Creating a Duration from milliseconds");
    TextualError.addMessage(it, "Argument 'milliseconds' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'milliseconds'", milliseconds);
    return Tried.Failure(it);
  }
  
  if (milliseconds < 0) {
    const it = TextualError.create("Creating a Duration from milliseconds");
    TextualError.addMessage(it, "Argument 'milliseconds' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'milliseconds'", milliseconds);
    return Tried.Failure(it);
  }
  
  if (milliseconds > MAXIMUM_MILLISECONDS) {
    const it = TextualError.create("Creating a Duration from milliseconds");
    TextualError.addMessage(it, "Argument 'milliseconds' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'milliseconds'", milliseconds);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_MILLISECONDS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(milliseconds));
};

export const fromSeconds = (seconds: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(seconds)) {
    const it = TextualError.create("Creating a Duration from seconds");
    TextualError.addMessage(it, "Argument 'seconds' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'seconds'", seconds);
    return Tried.Failure(it);
  }
  
  if (seconds < 0) {
    const it = TextualError.create("Creating a Duration from seconds");
    TextualError.addMessage(it, "Argument 'seconds' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'seconds'", seconds);
    return Tried.Failure(it);
  }
  
  if (seconds > MAXIMUM_SECONDS) {
    const it = TextualError.create("Creating a Duration from seconds");
    TextualError.addMessage(it, "Argument 'seconds' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'seconds'", seconds);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_SECONDS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(seconds * MILLISECONDS_PER_SECOND));
};

export const fromMinutes = (minutes: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(minutes)) {
    const it = TextualError.create("Creating a Duration from minutes");
    TextualError.addMessage(it, "Argument 'minutes' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'minutes'", minutes);
    return Tried.Failure(it);
  }
  
  if (minutes < 0) {
    const it = TextualError.create("Creating a Duration from minutes");
    TextualError.addMessage(it, "Argument 'minutes' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'minutes'", minutes);
    return Tried.Failure(it);
  }
  
  if (minutes > MAXIMUM_MINUTES) {
    const it = TextualError.create("Creating a Duration from minutes");
    TextualError.addMessage(it, "Argument 'minutes' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'minutes'", minutes);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_MINUTES);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(minutes * MILLISECONDS_PER_MINUTE));
};

export const fromHours = (hours: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(hours)) {
    const it = TextualError.create("Creating a Duration from hours");
    TextualError.addMessage(it, "Argument 'hours' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'hours'", hours);
    return Tried.Failure(it);
  }
  
  if (hours < 0) {
    const it = TextualError.create("Creating a Duration from hours");
    TextualError.addMessage(it, "Argument 'hours' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'hours'", hours);
    return Tried.Failure(it);
  }
  
  if (hours > MAXIMUM_HOURS) {
    const it = TextualError.create("Creating a Duration from hours");
    TextualError.addMessage(it, "Argument 'hours' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'hours'", hours);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_HOURS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(hours * MILLISECONDS_PER_HOUR));
};

export const fromDays = (days: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(days)) {
    const it = TextualError.create("Creating a Duration from days");
    TextualError.addMessage(it, "Argument 'days' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'days'", days);
    return Tried.Failure(it);
  }
  
  if (days < 0) {
    const it = TextualError.create("Creating a Duration from days");
    TextualError.addMessage(it, "Argument 'days' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'days'", days);
    return Tried.Failure(it);
  }
  
  if (days > MAXIMUM_DAYS) {
    const it = TextualError.create("Creating a Duration from days");
    TextualError.addMessage(it, "Argument 'days' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'days'", days);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_DAYS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(days * MILLISECONDS_PER_DAY));
};

export const fromWeeks = (weeks: number): Tried<Duration, TextualError> => {
  if (!Number.isInteger(weeks)) {
    const it = TextualError.create("Creating a Duration from weeks");
    TextualError.addMessage(it, "Argument 'weeks' is not an integer")
    TextualError.addNumberAttachment(it, "Argument 'weeks'", weeks);
    return Tried.Failure(it);
  }
  
  if (weeks < 0) {
    const it = TextualError.create("Creating a Duration from weeks");
    TextualError.addMessage(it, "Argument 'weeks' is negative: This Duration type only suuports representing positive durations")
    TextualError.addNumberAttachment(it, "Argument 'weeks'", weeks);
    return Tried.Failure(it);
  }
  
  if (weeks > MAXIMUM_WEEKS) {
    const it = TextualError.create("Creating a Duration from weeks");
    TextualError.addMessage(it, "Argument 'weeks' is greater than maximum value")
    TextualError.addNumberAttachment(it, "Argument 'weeks'", weeks);
    TextualError.addNumberAttachment(it, "Maximum value",  MAXIMUM_WEEKS);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(weeks * MILLISECONDS_PER_WEEK));
};

export const fromMillisecondsOrThrow = (millseconds: number): Duration => {
  const it = fromMilliseconds(millseconds);
  if (Tried.isSuccess(it)) {
    return Tried.value(it);
  } else {
    throw new Error(TextualError.prettyPrint(Tried.error(it)));
  }
};

export const zero = (): Duration => {
  return construct(0);
};

export const isZero = (it: Duration): boolean => {
  return toTotalMilliseconds(it) === 0;
};

export const toTotalMilliseconds = (it: Duration): number => {
  return it;
};

export const toTotaMinutes = (it: Duration): number => {
  return Math.floor(toTotalMilliseconds(it) / MILLISECONDS_PER_MINUTE);
};

export const minusOrZero = (it: Duration, rhs: Duration): Duration => {
  if (toTotalMilliseconds(it) > toTotalMilliseconds(rhs)) {
    return construct(toTotalMilliseconds(it) - toTotalMilliseconds(rhs));
  } else {
    return zero();
  }
};

export const plusOrMax = (it: Duration, rhs: Duration): Duration => {
  const result = toTotalMilliseconds(it) + toTotalMilliseconds(rhs);
  if (result <= MAXIMUM_MILLISECONDS) {
    return construct(result);
  } else {
    return construct(MAXIMUM_MILLISECONDS);
  }
};

export const isEqualTo = (lhs: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(lhs) === toTotalMilliseconds(rhs);
};

export const isLongerThan = (it: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(it) > toTotalMilliseconds(rhs);
};

export const isLongerThanOrEqualTo = (it: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(it) >= toTotalMilliseconds(rhs);
};

export const isShorterThan = (it: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(it) < toTotalMilliseconds(rhs);
};

export const isShorterThanOrEqualTo = (it: Duration, rhs: Duration): boolean => {
  return toTotalMilliseconds(it) <= toTotalMilliseconds(rhs);
};

export const min = (lhs: Duration, rhs: Duration): Duration => {
  return isShorterThan(lhs, rhs) ? lhs : rhs;
};

export const max = (lhs: Duration, rhs: Duration): Duration => {
  return isLongerThan(lhs, rhs) ? lhs : rhs;
};

export const toString = (it: Duration): string => {
  const parts: string[] = [];

  let totalMilliseconds = toTotalMilliseconds(it);
  
  const totalDays = Math.floor(totalMilliseconds / MILLISECONDS_PER_DAY);
  totalMilliseconds %= MILLISECONDS_PER_DAY;
  parts.push(`${totalDays} D`);
  
  const totalHours = Math.floor(totalMilliseconds / MILLISECONDS_PER_HOUR);
  totalMilliseconds %= MILLISECONDS_PER_HOUR;
  parts.push(`${totalHours} H`);
  
  const totalMinutes = Math.floor(totalMilliseconds / MILLISECONDS_PER_MINUTE);
  totalMilliseconds %= MILLISECONDS_PER_MINUTE;
  parts.push(`${totalMinutes} M`);
  
  return parts.join(' ');
};

export const toString2 = (it: Duration): string => {
  const milliseconds = toTotalMilliseconds(it);

  if (milliseconds === 0) {
    return '0s';
  }

  const totalSeconds = Math.floor(milliseconds / 1000);
  const totalMinutes = Math.floor(totalSeconds / 60);
  const totalHours = Math.floor(totalMinutes / 60);
  const totalDays = Math.floor(totalHours / 24);

  const seconds = totalSeconds % 60;
  const minutes = totalMinutes % 60;
  const hours = totalHours % 24;
  const days = totalDays;

  const parts: string[] = [];

  if (days > 0) {
    parts.push(`${days}d`);
  }
  if (hours > 0) {
    parts.push(`${hours}h`);
  }
  if (minutes > 0) {
    parts.push(`${minutes}m`);
  }
  if (seconds > 0 || parts.length === 0) {
    parts.push(`${seconds}s`);
  }

  return parts.join(' ');
};

export const Duration = {
  MAXIMUM_MILLISECONDS,
  MAXIMUM_SECONDS,
  MAXIMUM_MINUTES,
  MAXIMUM_HOURS,
  MILLISECONDS_PER_SECOND,
  MILLISECONDS_PER_MINUTE,
  MILLISECONDS_PER_HOUR,
  MILLISECONDS_PER_DAY,
  MILLISECONDS_PER_WEEK,
  fromMilliseconds,
  fromMinutes,
  fromSeconds,
  fromHours,
  fromDays,
  fromWeeks,
  zero,
  toTotalMilliseconds,
  toTotaMinutes,
  isZero,
  minusOrZero,
  plusOrMax,
  isEqualTo,
  isLongerThan,
  isLongerThanOrEqualTo,
  isShorterThan,
  isShorterThanOrEqualTo,
  min,
  max,
  toString,
  toString2,
  fromMillisecondsOrThrow,
};

File: discipline4/app/src/main/java/com/example/app/chronic/Duration.kt
package com.example.app

import arrow.core.Either
import arrow.core.raise.either
import arrow.core.right
import kotlin.math.floor
import kotlin.math.min
import kotlin.math.max
import com.example.app.TextualError

@JvmInline
public value class Duration private constructor(val milliseconds: Long) {
  companion object {
    const val MILLISECONDS_PER_SECOND = 1000L
    const val MILLISECONDS_PER_MINUTE = MILLISECONDS_PER_SECOND * 60
    const val MILLISECONDS_PER_HOUR = MILLISECONDS_PER_MINUTE * 60
    const val MILLISECONDS_PER_DAY = MILLISECONDS_PER_HOUR * 24
    const val MILLISECONDS_PER_WEEK = MILLISECONDS_PER_DAY * 7

    const val MAXIMUM_MILLISECONDS = Long.MAX_VALUE
    const val MAXIMUM_SECONDS = MAXIMUM_MILLISECONDS / MILLISECONDS_PER_SECOND
    const val MAXIMUM_MINUTES = MAXIMUM_SECONDS / 60
    const val MAXIMUM_HOURS = MAXIMUM_MINUTES / 60
    const val MAXIMUM_DAYS = MAXIMUM_HOURS / 24
    const val MAXIMUM_WEEKS = MAXIMUM_DAYS / 7

    /**
     * Creates a Duration from milliseconds
     */
    fun fromMilliseconds(milliseconds: Long): Either<TextualError, Duration> = either {
      if (milliseconds < 0) {
        TextualError.create("Creating a Duration from milliseconds")
          .addMessage("Argument 'milliseconds' is negative: This Duration type only supports representing positive durations")
          .addNumberAttachment("Argument 'milliseconds'", milliseconds.toDouble())
          .left()
          .bind()
      }
      
      if (milliseconds > MAXIMUM_MILLISECONDS) {
        TextualError.create("Creating a Duration from milliseconds")
          .addMessage("Argument 'milliseconds' is greater than maximum value")
          .addNumberAttachment("Argument 'milliseconds'", milliseconds.toDouble())
          .addNumberAttachment("Maximum value", MAXIMUM_MILLISECONDS.toDouble())
          .left()
          .bind()
      }

      Duration(milliseconds).right().bind()
    }

    /**
     * Creates a Duration from seconds
     */
    fun fromSeconds(seconds: Long): Either<TextualError, Duration> = either {
      if (seconds < 0) {
        TextualError.create("Creating a Duration from seconds")
          .addMessage("Argument 'seconds' is negative: This Duration type only supports representing positive durations")
          .addNumberAttachment("Argument 'seconds'", seconds.toDouble())
          .left()
          .bind()
      }
      
      if (seconds > MAXIMUM_SECONDS) {
        TextualError.create("Creating a Duration from seconds")
          .addMessage("Argument 'seconds' is greater than maximum value")
          .addNumberAttachment("Argument 'seconds'", seconds.toDouble())
          .addNumberAttachment("Maximum value", MAXIMUM_SECONDS.toDouble())
          .left()
          .bind()
      }

      Duration(seconds * MILLISECONDS_PER_SECOND).right().bind()
    }

    /**
     * Creates a Duration from minutes
     */
    fun fromMinutes(minutes: Long): Either<TextualError, Duration> = either {
      if (minutes < 0) {
        TextualError.create("Creating a Duration from minutes")
          .addMessage("Argument 'minutes' is negative: This Duration type only supports representing positive durations")
          .addNumberAttachment("Argument 'minutes'", minutes.toDouble())
          .left()
          .bind()
      }
      
      if (minutes > MAXIMUM_MINUTES) {
        TextualError.create("Creating a Duration from minutes")
          .addMessage("Argument 'minutes' is greater than maximum value")
          .addNumberAttachment("Argument 'minutes'", minutes.toDouble())
          .addNumberAttachment("Maximum value", MAXIMUM_MINUTES.toDouble())
          .left()
          .bind()
      }

      Duration(minutes * MILLISECONDS_PER_MINUTE).right().bind()
    }

    /**
     * Creates a Duration from hours
     */
    fun fromHours(hours: Long): Either<TextualError, Duration> = either {
      if (hours < 0) {
        TextualError.create("Creating a Duration from hours")
          .addMessage("Argument 'hours' is negative: This Duration type only supports representing positive durations")
          .addNumberAttachment("Argument 'hours'", hours.toDouble())
          .left()
          .bind()
      }
      
      if (hours > MAXIMUM_HOURS) {
        TextualError.create("Creating a Duration from hours")
          .addMessage("Argument 'hours' is greater than maximum value")
          .addNumberAttachment("Argument 'hours'", hours.toDouble())
          .addNumberAttachment("Maximum value", MAXIMUM_HOURS.toDouble())
          .left()
          .bind()
      }

      Duration(hours * MILLISECONDS_PER_HOUR).right().bind()
    }

    /**
     * Creates a Duration from days
     */
    fun fromDays(days: Long): Either<TextualError, Duration> = either {
      if (days < 0) {
        TextualError.create("Creating a Duration from days")
          .addMessage("Argument 'days' is negative: This Duration type only supports representing positive durations")
          .addNumberAttachment("Argument 'days'", days.toDouble())
          .left()
          .bind()
      }
      
      if (days > MAXIMUM_DAYS) {
        TextualError.create("Creating a Duration from days")
          .addMessage("Argument 'days' is greater than maximum value")
          .addNumberAttachment("Argument 'days'", days.toDouble())
          .addNumberAttachment("Maximum value", MAXIMUM_DAYS.toDouble())
          .left()
          .bind()
      }

      Duration(days * MILLISECONDS_PER_DAY).right().bind()
    }

    /**
     * Creates a Duration from weeks
     */
    fun fromWeeks(weeks: Long): Either<TextualError, Duration> = either {
      if (weeks < 0) {
        TextualError.create("Creating a Duration from weeks")
          .addMessage("Argument 'weeks' is negative: This Duration type only supports representing positive durations")
          .addNumberAttachment("Argument 'weeks'", weeks.toDouble())
          .left()
          .bind()
      }
      
      if (weeks > MAXIMUM_WEEKS) {
        TextualError.create("Creating a Duration from weeks")
          .addMessage("Argument 'weeks' is greater than maximum value")
          .addNumberAttachment("Argument 'weeks'", weeks.toDouble())
          .addNumberAttachment("Maximum value", MAXIMUM_WEEKS.toDouble())
          .left()
          .bind()
      }

      Duration(weeks * MILLISECONDS_PER_WEEK).right().bind()
    }

    /**
     * Creates a Duration from milliseconds or throws an exception
     */
    fun fromMillisecondsOrThrow(milliseconds: Long): Duration {
      return fromMilliseconds(milliseconds).fold(
        ifLeft = { error -> throw IllegalStateException(error.prettyPrint()) },
        ifRight = { it }
      )
    }

    /**
     * Returns a zero duration
     */
    fun zero(): Duration = Duration(0)

    private fun construct(milliseconds: Long): Duration = Duration(milliseconds)
  }

  /**
   * Returns the total milliseconds of this duration
   */
  fun toTotalMilliseconds(): Long = milliseconds

  /**
   * Returns the total minutes (floor) of this duration
   */
  fun toTotalMinutes(): Long = milliseconds / MILLISECONDS_PER_MINUTE

  /**
   * Returns true if this duration is zero
   */
  fun isZero(): Boolean = milliseconds == 0L

  /**
   * Subtracts another duration, returns zero if result would be negative
   */
  fun minusOrZero(rhs: Duration): Duration {
    return if (milliseconds > rhs.milliseconds) {
      Duration(milliseconds - rhs.milliseconds)
    } else {
      zero()
    }
  }

  /**
   * Adds another duration, caps at maximum value
   */
  fun plusOrMax(rhs: Duration): Duration {
    val result = milliseconds + rhs.milliseconds
    return if (result <= MAXIMUM_MILLISECONDS) {
      Duration(result)
    } else {
      Duration(MAXIMUM_MILLISECONDS)
    }
  }

  /**
   * Returns true if this duration is equal to another
   */
  fun isEqualTo(rhs: Duration): Boolean = milliseconds == rhs.milliseconds

  /**
   * Returns true if this duration is longer than another
   */
  fun isLongerThan(rhs: Duration): Boolean = milliseconds > rhs.milliseconds

  /**
   * Returns true if this duration is longer than or equal to another
   */
  fun isLongerThanOrEqualTo(rhs: Duration): Boolean = milliseconds >= rhs.milliseconds

  /**
   * Returns true if this duration is shorter than another
   */
  fun isShorterThan(rhs: Duration): Boolean = milliseconds < rhs.milliseconds

  /**
   * Returns true if this duration is shorter than or equal to another
   */
  fun isShorterThanOrEqualTo(rhs: Duration): Boolean = milliseconds <= rhs.milliseconds

  /**
   * Returns a human-readable string representation (e.g., "5d 3h 2m")
   */
  override fun toString(): String {
    var remainingMs = milliseconds
    val parts = mutableListOf<String>()
    
    val days = remainingMs / MILLISECONDS_PER_DAY
    remainingMs %= MILLISECONDS_PER_DAY
    if (days > 0) parts.add("${days}d")
    
    val hours = remainingMs / MILLISECONDS_PER_HOUR
    remainingMs %= MILLISECONDS_PER_HOUR
    if (hours > 0) parts.add("${hours}h")
    
    val minutes = remainingMs / MILLISECONDS_PER_MINUTE
    remainingMs %= MILLISECONDS_PER_MINUTE
    if (minutes > 0) parts.add("${minutes}m")
    
    val seconds = remainingMs / MILLISECONDS_PER_SECOND
    if (seconds > 0 || parts.isEmpty()) parts.add("${seconds}s")
    
    return parts.joinToString(" ")
  }

  /**
   * Alternative string representation with full units (e.g., "5 Days 3 Hours 2 Minutes")
   */
  fun toLongString(): String {
    var remainingMs = milliseconds
    val parts = mutableListOf<String>()
    
    val days = remainingMs / MILLISECONDS_PER_DAY
    remainingMs %= MILLISECONDS_PER_DAY
    if (days > 0) parts.add("${days} D")
    
    val hours = remainingMs / MILLISECONDS_PER_HOUR
    remainingMs %= MILLISECONDS_PER_HOUR
    if (hours > 0) parts.add("${hours} H")
    
    val minutes = remainingMs / MILLISECONDS_PER_MINUTE
    remainingMs %= MILLISECONDS_PER_MINUTE
    if (minutes > 0) parts.add("${minutes} M")
    
    return parts.joinToString(" ")
  }

  operator fun plus(other: Duration): Duration = plusOrMax(other)
  operator fun minus(other: Duration): Duration = minusOrZero(other)
  operator fun compareTo(other: Duration): Int = milliseconds.compareTo(other.milliseconds)
}

// Extension functions for easier usage
fun Duration.totalSeconds(): Long = this.milliseconds / Duration.MILLISECONDS_PER_SECOND
fun Duration.totalMinutes(): Long = this.milliseconds / Duration.MILLISECONDS_PER_MINUTE
fun Duration.totalHours(): Long = this.milliseconds / Duration.MILLISECONDS_PER_HOUR
fun Duration.totalDays(): Long = this.milliseconds / Duration.MILLISECONDS_PER_DAY
fun Duration.totalWeeks(): Long = this.milliseconds / Duration.MILLISECONDS_PER_WEEK

// Convenience functions for creating durations
fun milliseconds(value: Long): Either<TextualError, Duration> = Duration.fromMilliseconds(value)
fun seconds(value: Long): Either<TextualError, Duration> = Duration.fromSeconds(value)
fun minutes(value: Long): Either<TextualError, Duration> = Duration.fromMinutes(value)
fun hours(value: Long): Either<TextualError, Duration> = Duration.fromHours(value)
fun days(value: Long): Either<TextualError, Duration> = Duration.fromDays(value)
fun weeks(value: Long): Either<TextualError, Duration> = Duration.fromWeeks(value)

// Infix functions for natural language operations
infix fun Duration.plusOrMax(other: Duration): Duration = this.plusOrMax(other)
infix fun Duration.minusOrZero(other: Duration): Duration = this.minusOrZero(other)
infix fun Duration.isLongerThan(other: Duration): Boolean = this.isLongerThan(other)
infix fun Duration.isShorterThan(other: Duration): Boolean = this.isShorterThan(other)

// Top-level functions for min/max
fun min(lhs: Duration, rhs: Duration): Duration = if (lhs.isShorterThan(rhs)) lhs else rhs
fun max(lhs: Duration, rhs: Duration): Duration = if (lhs.isLongerThan(rhs)) lhs else rhs